Convert all our non-xfailed REPL tests to lit tests

vedantk opened the pull request:
This converts all of our REPL tests which use pexpect-backed routines to lit tests, which are more reliable.

There is one exception: I've left TestREPLBreakpoints.py alone because it's still x-failed. We'll need to look into that one separately.

vedantk committed:
[REPL] test: Convert TestREPLRecursiveClass.py to a lit test

vedantk committed:
[REPL] test: Convert TestREPLTypeLookup.py to a lit test

vedantk committed:
[REPL] test: Convert TestREPLStructs.py to a lit test

vedantk committed:
[REPL] test: Convert TestREPLQuickLookObject.py to a lit test

vedantk committed:
[REPL] test: Convert TestREPLClasses.py to a lit test

vedantk committed:
[REPL] test: Add a lit test for REPL breakpoints

vedantk committed:
[REPL] Consolidate 6 tests into one lit test

vedantk committed:
[REPL] test: Convert TestPOUnwrapping.py to a lit test

vedantk commented:
@swift-ci test

dcci reviewed (changes requested):
I think you might want to make these `REQUIRES: Darwin` as we haven't finished to add the CMake support to Linux for the repl. I think these are going to fail on the bot. Other than that, it looks fine.

vedantk commented:
@dcci gotcha, I’ll look into fleshing out the Linux cmake support tomorrow. I’d rather not land this without deleting the pexpect tests that need to be replaced.

jimingham commented on lit/SwiftExpr/BreakpointREPL.test:
```diff
@@ -0,0 +1,13 @@
+  return 0
+}
+
+:b foo
```
You switched this from a file & line breakpoint in the original test to a function breakpoint.  It would be fine to add a function name test but you shouldn't remove the test for file & line breakpoints.

jimingham replied:
That's the hard part, after all - making sure that subsequent REPL invocations keep a useful line number mapping...

vedantk replied:
This is a new test, it's not meant to replace anything (I mentioned this in the PR description). It's just a basic sanity check.

jimingham commented on lit/SwiftExpr/RecursiveClass.test:
```diff
@@ -0,0 +1,43 @@
+// CHECK-NEXT:   x = "Hello World"
+// CHECK-NEXT: }
+
+// Also test the more basic non-recursive case.
```
The Recursive class test was testing a particular tricky bit of the REPL - that you could use an optional of a class in it's definition.  Combining it with the basic "can we define classes" tests seems confusing to me.  Now if I want to add another test of class behavior where would I put it?  We should either keep this one narrowly focused and make a generic class behavior test (by just changing the name of the test...) or make another test for collecting more general class behavior.

vedantk replied:
Sure, I'll split it back out.

jimingham commented on lit/SwiftExpr/TypeLookup.test:
```diff
@@ -0,0 +1,13 @@
+// Test that type lookup chooses the right language.
```
Can we change the name of this test.  You didn't do this, but...  The point of the test is that, having imported Foundation, we can lookup a type from it.  So it's really a test of importing Foundation, type lookup is really just the means.  It would actually be better to have the test try to use Foundation...  And testing the functionality of "type lookup" which is a straight-up lldb command isn't best done in a REPL test...

vedantk replied:
I'll find a better name. Maybe ImportFoundation.test.

jimingham commented on lit/SwiftExpr/SimpleExpressions.test:
```diff
@@ -0,0 +1,71 @@
+// Test that we can define and use basic functions/expressions in the REPL.
```
This one does pretty much everything TestREPLBasic does.  Add a simple variable assignment, and then you should be able to delete that test as well.

vedantk replied:
Will do, thanks!

jimingham reviewed (changes requested):
Why is the directory for these tests called SwiftExpr?  That makes it sound like it is testing the swift expression parser, but it's actually testing REPL behavior.  Some of that tests generic expression parser behavior, but the REPL really only does a small subset of the full expression parser's jobs.  Be better to call this SwiftREPL.

vedantk committed:
[REPL] test: Clean up the test layout

- Split some of the testing in RecursiveClass.test out into Class.test.
- Rename the SwiftExpr test directory to SwiftREPL.
- Rename TypeLookup.test to ImportFoundation.test.

vedantk committed:
[REPL] test: Convert TestREPLBasic.py to a lit test

vedantk commented:
@jimingham Thanks for your comments, I believe I've addressed them now.

I still need to get things working on Linux before we can kick off testing.

vedantk committed:
[REPL] Build the dummy repl on Linux for testing

vedantk commented:
@swift-ci test

vedantk commented:
@swift-ci test

vedantk commented:
Great, tests look good on macOS/Linux now. Is this fine to merge?

jimingham commented:
This is fine.  If you add a file & line test to the breakpoints test, then you can delete TestREPLBreakpoints.py which gets us further to no pexpect based tests, but you can do that as a separate commit.

dcci commented:
The plumbing is correct, I only am uncertain about a macro, but if that works (and you tested) please go ahead

dcci commented on tools/repl/swift/main.c:
```diff
@@ -12,9 +12,10 @@
 #endif
 
-#ifdef __APPLE__
+#ifdef __gnu_linux__
```
Linux has kind aleph-zero different ways of telling to the preprocessor "am I running Linux", so we might change this in the future.

vedantk replied:
This was the first macro I found in clang that can be used to identify linux targets.

But, yes, we'll absolutely need to rethink this later.

dcci commented on tools/repl/swift/main.c:
```diff
@@ -30,6 +31,8 @@ int main() {
 #ifdef __APPLE__
   // Force loading of libswiftCore.dylib, which is not linked at build time.
   dlopen("@rpath/libswiftCore.dylib", RTLD_LAZY);
+#else
```
`#elif Linux` ?

vedantk commented:
Testing on rad-linux looks good. The macro guards in the dummy repl's main.c aren't great, but we can iterate on it as needed.
